Schema class inheritance support

jturkel opened the pull request:
This is a first stab at support for inheriting `GraphQL::Schema` settings (i.e. fixing #1627) e.g.

```ruby
class BaseSchema < GraphQL::Schema
  use GraphQL::Tracing::NewRelicTracing
  use GraphQL::Execution::Interpreter
  use GraphQL::Analysis::AST
  max_complexity 5000
end

class ServiceSchema < BaseSchema
  # ...
end

class AdminSchema < BaseSchema
  # ...
end
```

I took a pretty naive approach to the implementation and just leveraged the existing `find_inherited_method` method from`GraphQL::Schema::Member::BaseDSLMethods` to walk up the superclass chain looking for non-null setting values. It's a bit tedious and it's error prone as new settings are added but it works. There appears to be more generic frameworks for defining inheritable settings in `GraphQL::Schema::Member::AcceptsDefinition` and `GraphQL::Define::InstanceDefinable` and I briefly considered a simplified port of ActiveSupport's [class_attribute](https://api.rubyonrails.org/classes/Class.html#method-i-class_attribute) but I'm assuming these weren't leveraged here for a good reason? I still need to finish the tests but I was hoping to get some high level feedback on the approach before I spend too much more time on it. Thanks!

rmosolgo commented:
:+1: This is looking great, thanks for taking a stab at it! 

If you want to add some DRY-er options for looking up inherited values, I'm totally open to it. I didn't work too hard on it because I was focusing on other things, but I agree it could be better. 

I'd rather stay away from the `.define`/`AcceptsDefinitions` stuff because that's all for backwards-compatibility, and I'd like to make a clean deletion of it in GraphQL-Ruby 2.0. We shouldn't add a dependency on ActiveSupport, but if you want to add some methods inspired by that, I'm open to it! 

Anything else you need from me to make this PR not a "draft"?

jturkel commented:
That's great context and the answer I was hoping for! Thanks. I'll work on getting this ready for a real review.

jturkel commented:
This should be ready for a few real review now. The biggest change since the draft was I changed the behavior of collection settings to lazily merge inherited values to make it consistent with other places like [GraphQL::Schema::Object#interfaces](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/schema/object.rb#L90) e.g. 

```ruby
class BaseSchema < GraphQL::Schema
  
end

class ServiceSchema < BaseSchema
  rescue_from ServiceException, ->(err) { } 
end

BaseSchema.rescue_from StandardError, ->(err) { } 

ServiceSchema.rescues
# { StandardError => ..., ServiceException => ... }
``` 
As part of this I changed the `orphan_types` and `directives` methods to append values rather than replace them to make the behavior of calling them multiple times more consistent e.g. 

```ruby
class BaseSchema < GraphQL::Schema
  orphan_types Foo
end

class ServiceSchema < BaseSchema
  orphan_types Bar
end
```
now results in the same values for `ServiceSchema.orphan_types` as:

```
class ServiceSchema < GraphQL::Schema
  orphan_types Foo
  orphan_types Bar
end
```

A few other changes worth calling out since the WIP:

* I renamed `find_inherited_method` to `find_inherited_value` to make it clearer that these methods are returning values rather than a [Method](https://ruby-doc.org/core-2.5.3/Method.html). I also made `nil` the default for `default_value` since that is the most common behavior and hopefully a reasonable default. Neither of these changes is essential so I'm happy to roll them back if you're not a fan.
* I updated several of the tests to work on subclasses of `GraphQL::Schema` to avoid mutating global settings that had to be rolled back in `after` blocks.

I punted on trying to factor out some of the common patterns in dealing with these inheritable settings. I think it probably makes sense to wait to see how things shake out in 2.0 after the legacy classes are removed.

rmosolgo commented on spec/graphql/analysis/analyze_query_spec.rb:
```diff
@@ -119,14 +121,7 @@ def call(memo, visit_type, irep_node)
-      after do
-        Dummy::Schema.query_analyzers.clear
-        Dummy::Schema.query_analyzers.push(*@previous_query_analyzers)
+        schema.query_analyzer(variable_accessor)
```
👌

rmosolgo commented on spec/graphql/analysis/ast/max_query_complexity_spec.rb:
```diff
@@ -2,14 +2,7 @@
-    Dummy::Schema.max_complexity = @prev_max_complexity
-  end
-
+  let(:schema) { Class.new(Dummy::Schema) }
```
👌
